### PR TITLE
Issue #3268943 by SV: As a LU I want to be able to see the list of the followers & following users

### PR DESCRIPTION
**Pull request by volodymyr-sydor-lemberg**
## Note
- Depend on [related PR](https://github.com/goalgorilla/open_social/pull/2778) and should be reviewed after merging that PR

## Problem
- LU should be able to see the list of the users another user follows
- LU should be able to see the list of users following another user
- LU should be able to follow other users from their profile teaser

## Solution

-  On the user’s profile it should be possible to see the list of people that user follow 
   - there should be a CTA leading to the list
   - there should be the number of users on the list specified
- On the user’s profile it should be possible to see the list of followers
   - there should be a CTA leading to the list
   - there should be the number of followers on the list specified
 - On the teaser user profile it should be possible to follow that user
   - there should be a Follow button
   - there should be numbers of followers/following users
   
## Issue tracker
- https://www.drupal.org/project/social/issues/3268943
- https://getopensocial.atlassian.net/browse/YANG-6957
- https://getopensocial.atlassian.net/browse/YANG-6958
- https://getopensocial.atlassian.net/browse/YANG-7265

## How to test
- [ ] Using  the latest version of Open Social with the social_follow_user module enabled
- [ ] As a LU (first users)
- [ ] Try to open /all-member page
- [ ] And click follow button for a few users
- [ ] Next, got to your profile page 
- [ ] And you should see the increased number of "following" users
- [ ] After click on the "following" link  you will be redirected to the "following" users page
- [ ] And you should see the list of "following" users 
- [ ] Login as another user (second user)
- [ ] And find "first" LU and click on the follow button 
- [ ] Switch back to the "first" LU and number of "followers" users should be increased
- [ ] After click on the "followers" link  you will be redirected to the "followers" users page
- [ ] And you should see the list of "followers 

## Screenshots

Screenshot 2022-03-09 at 12 33 28
Screenshot 2022-03-09 at 18 44 52

## Release notes
 - Provides a list of following users & followers

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

File: modules/social_features/social_profile/modules/social_profile_preview/src/Service/SocialProfilePreviewHelper.php [hunk 1]
```diff
@@ -38,6 +38,8 @@ public function alter(
       }
 
       $variables['#attached']['library'][] = 'social_profile_preview/base';
+      $variables['#attached']['library'][] = 'social_follow_user/follow_user_counter';
+      $variables['#attached']['library'][] = 'social_follow_user/follow_user_btn';
```
Comment (OH17L) [inline modules/social_features/social_profile/modules/social_profile_preview/src/Service/SocialProfilePreviewHelper.php:42]: Check if the social_follow_user module is enabled.

File: modules/social_features/social_profile/modules/social_profile_preview/templates/profile--profile--preview.html.twig [hunk 2]
```diff
@@ -54,5 +54,10 @@
     </div>
   </div>
 
-  {{ content|without('field_profile_image', 'uid', 'flag_follow_user', 'private_message_link') }}
+  {{ content.field_profile_self_introduction }}
+
+  <div class='follow-user--counter'>
+    {{ content|without('field_profile_image', 'uid', 'flag_follow_user', 'private_message_link', 'field_profile_self_introduction') }}
+  </div>
+
```
Comment (OH17L) [inline modules/social_features/social_profile/modules/social_profile_preview/templates/profile--profile--preview.html.twig:62]:
```suggestion
  {% if content.followers or content.following %}
    <div class='follow-user--counter'>
      {{ content.followers }}
      {{ content.following }}
    </div>
  {% endif %}
  
  {{ content|without('field_profile_image', 'uid', 'flag_follow_user', 'private_message_link', 'field_profile_self_introduction', 'followers', 'following') }}
```

File: modules/social_features/social_follow_user/social_follow_user.module [hunk 3]
```diff
@@ -97,14 +97,20 @@ function social_follow_user_profile_view(
...
+  if ($entity->bundle() === 'profile') {
+
+    if ($display->getComponent('social_follow_user') !== NULL) {
+      $build['social_follow_user'] = [
+        '#theme' => 'followers',
+        '#flaggable' => $entity,
+      ];
+    }
+    if ($display->getComponent('social_following_user') !== NULL) {
+      $build['social_following_user'] = [
```
Comment (OH17L) [inline modules/social_features/social_follow_user/social_follow_user.module:109]: Define the field in the "social_follow_user_entity_extra_field_info" hook and make the following changes:
```suggestion
    if ($display->getComponent('following') !== NULL) {
      $build['following'] = [
```

File: modules/social_features/social_follow_user/social_follow_user.libraries.yml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+follow_user_counter:
+  css:
+    base:
+      assets/css/social_follow_user_counter.css: {}
+
+follow_user_btn:
```
Comment (OH17L) [inline modules/social_features/social_follow_user/social_follow_user.libraries.yml:6]:
```suggestion
button:
```

File: modules/social_features/social_follow_user/social_follow_user.libraries.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
...
+  css:
+    base:
+      assets/css/social_follow_user_btn.css: {}
```
Comment (OH17L) [inline modules/social_features/social_follow_user/social_follow_user.libraries.yml:9]:
```suggestion
      assets/css/social_follow_user.button.css: {}
```

File: modules/social_features/social_follow_user/social_follow_user.module [hunk 6]
```diff
@@ -117,8 +123,52 @@ function social_follow_user_theme(): array {
...
+
+/**
+ * Implements hook_preprocess_HOOK().
+ */
+function social_follow_user_preprocess_profile(array &$variables): void {
+  if ($variables['elements']['#view_mode'] === 'teaser') {
+    // Display related teaser tag for each user follow list.
+    switch (\Drupal::routeMatch()->getRouteName()) {
+      case 'view.followers_user.page_user_followers':
+        $variables['social_follower'] = TRUE;
```
Comment (OH17L) [inline modules/social_features/social_follow_user/social_follow_user.module:149]:
```suggestion
        $variables['followers'] = TRUE;
```

File: modules/social_features/social_follow_user/social_follow_user.module [hunk 7]
```diff
@@ -117,8 +123,52 @@ function social_follow_user_theme(): array {
...
+        break;
+
+      case 'view.following_users.page_following_users':
+        $variables['social_following'] = TRUE;
```
Comment (OH17L) [inline modules/social_features/social_follow_user/social_follow_user.module:153]:
```suggestion
        $variables['following'] = TRUE;
```

**Review by OH17L — CHANGES_REQUESTED** [review]
Add patches for themes to Composer's file and make improvements described below:

File: modules/social_features/social_follow_user/config/install/views.view.following_users.yml [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+    display_plugin: default
+    id: default
+    display_title: Default
+    position: 0
+    display_options:
+      access:
+        type: role
+        options:
+          role:
+            authenticated: authenticated
```
[the remark this answers is not in the document]
Comment (volodymyr-sydor-lemberg) [reply modules/social_features/social_follow_user/config/install/views.view.following_users.yml:33]: I've found an existing one and would suggest using it: `flag follow_user`

File: modules/social_features/social_follow_user/config/install/views.view.followers_user.yml [hunk 9]
```diff
@@ -0,0 +1,338 @@
...
+    display_plugin: default
+    id: default
+    display_title: Default
+    position: 0
+    display_options:
+      access:
+        type: role
+        options:
+          role:
+            authenticated: authenticated
```
[the remark this answers is not in the document]
Comment (volodymyr-sydor-lemberg) [reply modules/social_features/social_follow_user/config/install/views.view.followers_user.yml:33]: I've found an existing one and would suggest using it: `flag follow_user`

File: modules/social_features/social_follow_user/src/EventSubscriber/FlagEvents.php [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    }
+  }
+
+  /**
+   * Invalidates cache tags.
+   */
+  public function invalidateCaches(): void {
+    $this->cacheTagsInvalidator->invalidateTags([
+      'followers_user',
+      'following_user',
```
[the remark this answers is not in the document]
Comment (volodymyr-sydor-lemberg) [reply modules/social_features/social_follow_user/src/EventSubscriber/FlagEvents.php:105]: when one user click to follow/unfollow another then both counter should be "updated"

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by open-social-tugboat** [comment]
[Tugboat](https://tugboat.qa) has finished building the preview for this pull request!

Link:
- https://pr2850-q4fkve0k0rllornm8roqxahl1c3vwdpb.tugboat.qa


Dashboard:
- https://dashboard.tugboat.qa/6246cd107dc9604ba3cb47ea

**Comment by OH17L** [comment]
Replaced by PR #2867.

**Pull request closed without merging**